File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/classes/property_delegates/boolean_property_delegate.ts
/**
 * Delegate for parsing JSON boolean values (true/false).
 */

import { PropertyDelegate } from "./property_delegate.js";
import { JsonStreamParserController } from "../json_stream_parser.js";

export class BooleanPropertyDelegate extends PropertyDelegate {
    private buffer = "";
    private expectedValue: boolean | null = null;

    constructor(
        propertyPath: string,
        parserController: JsonStreamParserController,
        onComplete?: () => void,
    ) {
        super(propertyPath, parserController, onComplete);
    }

    addCharacter(character: string): void {
        // Check for delimiters that end the boolean
        if (character === "," || character === "}" || character === "]") {
            if (!this.isDone && this.expectedValue !== null) {
                this.isDone = true;
                this.parserController.completeProperty(
                    this.propertyPath,
                    this.expectedValue,
                );
                this.onComplete?.();
            }
            return;
        }

        // First character determines which boolean we're parsing
        if (this.buffer.length === 0) {
            if (character === "t") {
                this.expectedValue = true;
                this.buffer = "t";
            } else if (character === "f") {
                this.expectedValue = false;
                this.buffer = "f";
            }
            return;
        }

        // Continue building the buffer
        this.buffer += character;

        // Check if we've completed parsing the boolean
        if (this.expectedValue === true && this.buffer === "true") {
            this.isDone = true;
            this.parserController.completeProperty(this.propertyPath, true);
            this.onComplete?.();
        } else if (this.expectedValue === false && this.buffer === "false") {
            this.isDone = true;
            this.parserController.completeProperty(this.propertyPath, false);
            this.onComplete?.();
        }
    }
}

File: packages/json/llm-json-stream-python/src/llm_json_stream/property_delegates/boolean_property_delegate.py
"""Delegate for parsing JSON boolean values."""

from __future__ import annotations

from ..property_delegate import PropertyDelegate


class BooleanPropertyDelegate(PropertyDelegate):
    def __init__(self, property_path: str, parser_controller, on_complete=None) -> None:
        super().__init__(property_path, parser_controller, on_complete)
        self._buffer = ""
        self._expected_value: bool | None = None

    def add_character(self, character: str) -> None:
        if character in {",", "}", "]"}:
            if not self._is_done and self._expected_value is not None:
                self._is_done = True
                self._parser_controller.complete_property(
                    self._property_path, self._expected_value
                )
                if self._on_complete:
                    self._on_complete()
            return

        if not self._buffer:
            if character == "t":
                self._expected_value = True
                self._buffer = "t"
            elif character == "f":
                self._expected_value = False
                self._buffer = "f"
            return

        self._buffer += character
        if self._expected_value is True and self._buffer == "true":
            self._is_done = True
            self._parser_controller.complete_property(self._property_path, True)
            if self._on_complete:
                self._on_complete()
        elif self._expected_value is False and self._buffer == "false":
            self._is_done = True
            self._parser_controller.complete_property(self._property_path, False)
            if self._on_complete:
                self._on_complete()
